library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextParser.java

```diff
@@ -0,0 +1,513 @@
+      }
+      data[offset + i] &= 0x7F;
+    }
+  }
+
+  // revert bits order from LSB first to MSB first
+  private static void revertBytes(byte[] data, int offset, int len) {
+    for (int i = 0; i < len; ++i) {
+      data[offset + i] = (byte) (TAB_REVERT_BITS[data[offset + i] & 0xFF] & 0xFF);
+    }
+  }
+
+  private static int hamming84Decode(int d0) {
+    int a = HAMMING_8_4_TABLE[d0];
+    if ((a & 0x80) != 0) {
+      Log.w(TAG, "uncorrectable hamming error");
+    }
+    return a & 0x0F;
+  }
+
+  private static int hamming84Decode(int d0, int d1) {
+    /* convert two consecutive hamming 8/4 encoded bytes */
+    int a = HAMMING_8_4_TABLE[d0];
+    int b = HAMMING_8_4_TABLE[d1];
+    if ((a & 0x80) != 0 || (b & 0x80) != 0) {
+      Log.w(TAG, "uncorrectable hamming error");
+    }
+    return ((b & 0x0F) << 4) | (a & 0x0F);
+  }
+
+  private static void handleSpecialChars(byte[] data, int offset, int len, int lang) {
+    boolean mosaic = false;
+    for (int i = offset; i < offset + len; ++i) {
+      int c = data[i] & 0xFF;
+      if (c >= 0x20) {
+        int rep = SPECIAL_CHARS_IDX[c];
+        if (rep != 0) {
+          if (!mosaic || ((rep >= 0x40) && (rep <= 0x50))) {
+            data[i] = (byte) (SPECIAL_CHARS[lang][rep] & 0xFF);
+          }
+        }
+      } else {
+        mosaic = c >= 0x10;
+      }
+    }
+  }
+
+  static {
+    SPECIAL_CHARS_IDX = new int[256];
+    SPECIAL_CHARS_IDX[0x23] = 1;
+    SPECIAL_CHARS_IDX[0x24] = 2;
+    SPECIAL_CHARS_IDX[0x40] = 3;
+    SPECIAL_CHARS_IDX[0x5b] = 4;
+    SPECIAL_CHARS_IDX[0x5c] = 5;
+    SPECIAL_CHARS_IDX[0x5d] = 6;
+    SPECIAL_CHARS_IDX[0x5e] = 7;
+    SPECIAL_CHARS_IDX[0x5f] = 8;
+    SPECIAL_CHARS_IDX[0x60] = 9;
+    SPECIAL_CHARS_IDX[0x7b] = 10;
+    SPECIAL_CHARS_IDX[0x7c] = 11;
+    SPECIAL_CHARS_IDX[0x7d] = 12;
+    SPECIAL_CHARS_IDX[0x7e] = 13;
+  }
+
+  private final int magazineNumber;
+  private final int pageNumber;
+  private final TeletextPage page = new TeletextPage();
+  private final ArrayList<Cue> cues = new ArrayList<>();
+
+  /**
+   * Construct an instance for the given magazine number and page number.
+   *
+   * @param magazineNumber The id of the subtitle magazine carrying the subtitle to be parsed.
+   * @param pageNumber The id of the subtitle page carrying the subtitle to be parsed.
+   */
+  public DvbTeletextParser(int magazineNumber, int pageNumber) {
+    this.magazineNumber = magazineNumber;
+    this.pageNumber = pageNumber;
+  }
+
+  /** Resets the parser. */
+  public void reset() {
+    page.clear();
+    cues.clear();
+  }
+
+  /**
+   * Decodes a subtitling packet, returning a list of parsed {@link Cue}s.
+   *
+   * @param data The subtitling packet data to decode.
+   * @param size The limit in {@code data} at which to stop decoding.
+   * @return The parsed {@link Cue}s.
+   */
+  public List<Cue> decode(byte[] data, int size) {
+    int offset = 0;
+    int len = size;
+
+    int dataIdentifier = data[offset++] & 0xFF;
```

Comment:
This might be better with a `ParsableByteArray`? 

If you want to avoid the allocation you could use a scratch instance as a field, and reset it with `data` each time.

Reply:
Done in 69f49d5388fbdc6972d7ca7517171fad3617d4de